Workerbee2

patricksnape opened the pull request:
Now supports a number of new fields for computing statistics
about the job completion behaviour. Rather than booleans
for job completion and job claiming we now store timestamps.
We also prefer the OLDEST job rather than a random one.

We use auto-increment ids for the job ids and thus support easily
adding new jobs. A new unique_id field is maintained for the
previous use case of those who want to use it to store data
rather than the input_data field.

The previously optional data field becomes the input_data json
field which is used to store job data. This is matched
with the new output_data field (optional) for storing the results
of a job.

We also keep track of how long a successful job took.

Finally, we add two new binary exponential backoff strategies -
one for failures and one for 'busywaiting'. If busywait is
enabled, then when all jobs are completed the system will
sleep in increasing intervals whilst waiting for a new job
to appear. If a job fails, the system will sleep increasing
amounts until it recieves a non-failing job. There is also
a maximum number of retries and jobs larger than this value
will not be considered as available jobs.
## TODO
- [x] Update docstrings
- [x] Add some endpoints for statistics
- [ ] ~~Possibly add endpoints to reset certain jobs errors?~~ (later work)
- [x] Review by @jabooth
## Jabooth TODO additions
- [ ] add arrow conda dep (or conda-forge? Just pip for now?)
- [x] finalise terminology, update README
- [ ] add issues for expansion work
  - [ ] reset job errors
  - [ ] create experiment CLI

patricksnape committed:
Refactor create_db_handle to have more robust logic

And support an arbitrary logger.

patricksnape committed:
Add new workerbee2 behaviour

Now supports a number of new fields for computing statistics
about the job completion behaviour. Rather than booleans
for job completion and job claiming we now store timestamps.
We also prefer the OLDEST job rather than a random one.

We use auto-increment ids for the job ids and thus support easily
adding new jobs. A new unique_id field is maintained for the
previous use case of those who want to use it to store data
rather than the input_data field.

The previously optional data field becomes the input_data json
field which is used to store job data. This is matched
with the new output_data field (optional) for storing the results
of a job.

We also keep track of how long a successful job took.

Finally, we add two new binary exponential backoff strategies -
one for failures and one for 'busywaiting'. If busywait is
enabled, then when all jobs are completed the system will
sleep in increasing intervals whilst waiting for a new job
to appear. If a job fails, the system will sleep increasing
amounts until it recieves a non-failing job. There is also
a maximum number of retries and jobs larger than this value
will not be considered as available jobs.

patricksnape committed:
Fix izip_longest import on Python 2

jabooth commented on workerbee/postgresql.py:
```diff
@@ -28,257 +34,328 @@ def check_name(table_name):
+CREATE TABLE {tbl_name}(
+  id SERIAL PRIMARY KEY,
+  unique_id TEXT UNIQUE,
+  input_data json NOT NULL,
```
We could use `JSONB` rather than `JSON`. Main advantage being `JSONB` are equatable, and so, can be held under a unique constraint. Perhaps the schema should be 

``` diff
-  unique_id TEXT UNIQUE,
-  input_data json NOT NULL,
+  input_data JSONB NOT NULL UNIQUE,
-  output_data json,
+  output_data JSONB,
```

Then we can just drop the `unique_id` altogether here. Support is Postgres 9.4+, which I think is acceptable?

Argument: workerbee functions really should be deterministic/idempotent, so duplicate job payloads in a single experiment never make sense.

jabooth commented on workerbee/postgresql.py:
```diff
@@ -28,257 +34,328 @@ def check_name(table_name):
+
+    def __str__(self):
+        return '{} on {}@{}:{}'.format(self.database, self.user, self.host,
+                                       self.port)
```
Unsure how I feel about `DBConnectionInfo`. The change from:

``` py
db_info = dict(host='localhost', user='postgres', database='postgres', port=5432)
```

to

``` py
from workerbee.postgresql import DBConnectionInfo
db_info = DBConnectionInfo(host='localhost', user='postgres', database='postgres', port=5432)
```

Just adds that little extra friction in my mind. Do like the idea of gathering it into one kwarg through. If we went with the `dict` approach I'd obviously want the verification etc. 

How would you feel about sticking with a dict as the public API but build a `DBConnectionInfo` internally? Could be done at the top of `get_db_handle()`, a try/catch here and we could raise a nice clear error if the user provides a malformed dict?

jabooth commented on workerbee/postgresql.py:
```diff
@@ -28,257 +34,328 @@ def check_name(table_name):
+
+def get_db_handle(db_info=None, logger_name=DEFAULT_LOGGER):
+    if db_info is None:
+        db_info = DBConnectionInfo()
```
this would be

```
db_info = {}
```

jabooth commented on workerbee/postgresql.py:
```diff
@@ -28,257 +34,328 @@ def check_name(table_name):
+def get_db_handle(db_info=None, logger_name=DEFAULT_LOGGER):
+    if db_info is None:
+        db_info = DBConnectionInfo()
+
```
And then here something like

``` py
try:
    db_info_obj = DBConnectionInfo(**db_info)
catch:
    # the user provided a dict with keys other than 'host', 'port' etc etc
```

jabooth commented on workerbee/postgresql.py:
```diff
@@ -28,257 +34,328 @@ def check_name(table_name):
+
+        if not table_exists(db_handle, experiment_id):
+            raise ValueError("Table does not exist for experiment '{}'".format(
+                experiment_id))
```
This is undoubtedly a less ambiguous API, but I did at first try:

``` py
from workerbee.postgres import add_jobs
add_jobs('my_new_exp', [{'path': 'thepath'}])
```

before realising I needed:

``` py
from workerbee.postgres import create_experiment
create_experiment('my_new_exp', [{'path': 'thepath'}])
```

this then feels a little weird, because `create_experiment` is like `add_jobs` only it additionally creates the table (i.e. it still takes data). Re-running `create_experiment` with different job data silently fails to add the new job data, only raising a warning about the table already existing.

I think it might be better to commit further in one explicit/implicit direction here. Either:
1. You have to first call `create_experiment` to instantiate a table for the experiment. Does not take data to add. Calling again is a `ValueError`. `add_jobs`/`add_job` behave as they do now.
   - \+ Explicit and clear API.
   - \- Two function calls needed for common case of _make a new table and add some data_.
2. `create_experiment` goes away - `add_job`/`add_jobs` just implicitly create a table if needed.
   - \+ Optimises for common case - if the key matches, just ensure that exp exists and add jobs. What's the use of building a table and not putting jobs in it?
   - \- Implicit table creation - if you change `experiment_id` by accident you would be highly miffed.

jabooth replied:
Thinking a little further, maybe the best option is 1 (explicit), but with a keyword arg `autocreate=False`.

You could then do:

``` py
from workerbee.postgres import add_jobs
add_jobs('my_new_exp', [{'path': 'thepath'}], autocreate=True)
```

In the absence of that kwarg I could imagine this being a common pattern:

``` py
from workerbee.postgres import add_jobs, create_experiment
try:
    create_experiment('my_exp')
catch ASuitableException:
    pass
add_jobs('my_new_exp', [{'path': 'thepath'}])
```

(at least for the way I currently use workerbee). At least this way the user is explicitly opting in to the more loosey goosey behaviour (like `overwrite=True`)

jabooth commented:
@patricksnape On the whole love it! Will throw together a little stat tracker now and commit on top. Let me know your thoughts on API design stuff suggested above.

jabooth commented on workerbee/base.py:
```diff
@@ -1,6 +1,56 @@
+from functools import partial
+
+
+DEFAULT_LOGGER = logging.getLogger('workerbee_default')
```
'workerbee_default' looks a little grim in stdout:
<img width="1116" alt="screenshot 2016-08-20 15 04 09" src="https://cloud.githubusercontent.com/assets/1312873/17831683/6be1c7a0-66e7-11e6-9dfe-900a837253af.png">
Is the `_default` prefix a standard logging pattern? If not maybe `workerbee`?

patricksnape replied:
It can be whatever you want and you can choose the pattern with the formatter. If all that information offends you you can change the pattern.

patricksnape committed:
Change default logger - no name printed

patricksnape committed:
Remove unique_id field

Change default data type for input_data to JSONB (where possible)
which can have a UNIQUE constraint. Thus we don't require the
unique_id. On Postgres version < 9.4 we fall back to the JSON
type and the UNIQUE constraint is not enforced.

patricksnape committed:
Guard for table existing

patricksnape committed:
Allow passing password manually

jabooth committed:
added wbee command line utility for monitoring experiments

jabooth committed:
rename wbee to workerbee

patricksnape committed:
Remove the filesystem version

It's very slow and we've decided to commit on postgres as our
primary interface.

patricksnape committed:
Refactor postgres worker into class

Makes it much simpler to query the status of a jobset if all
the functionality is encapsulated by a class.

patricksnape committed:
Better error checking and unique constraint on input_data

Now support (partially) postgres down to 9.2 but can detect
below that and correctly raise a valueerror.

Also, rename database to dbname in the connection info

patricksnape committed:
Merge branch 'workerbee2' of github.com:jabooth/workerbee into workerbee2

patricksnape commented:
[`error: file '/home/travis/miniconda/conda-bld/work/bin/wbee' does not exist`](https://travis-ci.org/menpo/workerbee/jobs/154205872#L339)

jabooth committed:
updated CLI for latest changes

jabooth committed:
fix workerbee script location, add docopt dep

jabooth committed:
started work on fixed README

jabooth commented on workerbee/postgresql.py:
```diff
@@ -1,284 +1,394 @@
+        self.db_handle.run(SET_ROW_COMPLETED_BY_ID_QUERY.format(
+            tbl_name=self.jobset_id), parameters=params)
+
+    def get_uncompleted_unclaimed_job(self):
```
Maybe these methods should be underscored?
- gives us flexibility to change internal API
- simplifies tab completion suggestions down to public API only

patricksnape committed:
Lower arrow dep - add to meta.yaml

patricksnape committed:
Update inline documentation

patricksnape committed:
Add note about UNIQUE 'id' key

In the input_data, for PostgreSQL < 9.4, an `id` key is reserved
in the input_data in order to enforce the uniqueness constraint.

patricksnape committed:
Add UniqueInputDataConstraintError, catch_all_exceptions

Renamed JobFailed -> JobFailedError
Now when a non-unique job is added a custom exception is thrown.
setup_jobset returns a boolean about whether or not it succeeded.
The 'id' key must be unique, if it exists, on all postgres version
> 9.3.

jabooth committed:
fixes for edge cases

jabooth committed:
stats fixes for completed case

jabooth committed:
spacing improvements

jabooth committed:
more spacing improvements

patricksnape committed:
Expose catch_all_exceptions

patricksnape committed:
Update README

patricksnape committed:
Incorrect version spec for arrow

patricksnape committed:
Relax version constraints slightly

patricksnape committed:
Remove extra imports

jabooth committed:
Merge remote-tracking branch 'origin/workerbee2' into workerbee2

patricksnape committed:
Minor changes for cmd line util

patricksnape committed:
Silence logging for cmd line tool

patricksnape committed:
Add docs to UniqueInputDataConstraintError

jabooth commented:
So this is now ready to go - we just have some issues about how best to release it.

condaci assumes we can build for mac/linux/win python 2.7/3.4/3.5. Unfortunately there is no psycopg2 for Windows in the default channel. So we could:
1. build only for linux/mac in our conda channel
   - would show as failures for windows, but easy option
   - bit of a pain to manually build windows later
2. change condaci to allow sourcing from other channels (conda-forge)
   - minor pain
   - changes install instructions
3. build psycopg2 for menpo conda channel
   - lot of effort

The benefit of the new version is so great that I think we shouldn't get caught up too much in this.

jabooth committed:
fix for stats report on < PG 9.4

jabooth committed:
default to port 5432 if not provided in env

jabooth committed:
add UTF-8 encoding to stats (contains unicode)

patricksnape commented:
Jenkins rebuild this please

patricksnape commented:
Jenkins retest this please

jabooth commented:
Just going to get on with this for Linux only for now if necessary. It's such an improvement for workerbee.
